src/commands/fileHistory.ts

```diff
@@ -1,120 +1,156 @@
+    catch (error) {
+        logger.logError(error);
+    }
 }
 
-function onItemSelected(item: vscode.QuickPickItem, fileName: string, relativeFilePath: string) {
-    let itemPickList: vscode.QuickPickItem[] = [];
+
+async function onItemSelected(item: vscode.QuickPickItem, fileName: string, relativeFilePath: string) {
+    const commit = (<any>item).data;
+    const getThisFile = getFile(commit.sha1, relativeFilePath);
+    const getPreviousFile = getFile(commit.previousSha1, relativeFilePath);
+
+    const thisFile = await getThisFile;
+    const previousFile = await getPreviousFile;
+
+    const itemPickList: vscode.QuickPickItem[] = [];
     itemPickList.push({ label: 'View Change Log', description: 'Author, committer and message' });
-    itemPickList.push({ label: 'View File Contents', description: '' });
-    itemPickList.push({ label: 'Compare against workspace file', description: '' });
-    if (!(<any>item).isLast) {
+    if (thisFile !== 'fileUnavailable') {
+        itemPickList.push({ label: 'View File Contents', description: '' });
+    }
+    if (thisFile !== 'fileUnavailable' && fs.existsSync(fileName)) {
+        itemPickList.push({ label: 'Compare against workspace file', description: '' });
+    }
+    if (previousFile !== 'fileUnavailable' &&  thisFile !== 'fileUnavailable') {
         itemPickList.push({ label: 'Compare against previous version', description: '' });
     }
 
     vscode.window.showQuickPick(itemPickList, { placeHolder: item.label, matchOnDescription: true }).then(cmd => {
         if (!cmd) {
             return;
         }
-        let data = (<any>item).data;
-        if (cmd.label === itemPickList[0].label) {
+        const data = (<any>item).data;
+        if (cmd.label === 'View Change Log') {
             viewLog(data);
             return;
         }
-
-        if (cmd.label === itemPickList[1].label) {
-            viewFile(data, relativeFilePath);
+        if (cmd.label === 'View File Contents') {
+            viewFile(thisFile);
             return;
         }
-        if (cmd.label === itemPickList[2].label) {
-            launchFileCompareWithLocal(data, fileName, relativeFilePath);
+        if (cmd.label === 'Compare against workspace file') {
+            diffFiles(fileName, thisFile, commit.sha1, fileName, '');
             return;
         }
-        if (itemPickList.length > 3 && cmd.label === itemPickList[3].label) {
-            launchFileCompareWithPrevious(data, relativeFilePath);
+        if (cmd.label === 'Compare against previous version') {
+            diffFiles(fileName, previousFile, commit.previousSha1, thisFile, commit.sha1);
             return;
         }
     });
 }
 
-function viewFile(details: any, relativeFilePath: string) {
-    displayFile(details.sha1, relativeFilePath).then(() => { }, genericErrorHandler);
+async function viewFile(fileName: string) {
+    try {
+        vscode.workspace.openTextDocument(`[${path.basename(fileName)}](${fileName}`).then(document => {
```

Comment:
If we want a custom title, we could:  
- Create a temporary directory (use the temporary path to create a folder instead of a file)  
- Next create a file in this directory with the necessary name  
Now open this, and VS Code should ideally display just the file name (the name we want)